src/parallel.py

```diff
@@ -0,0 +1,188 @@
+
+    Because called methods in this function must also open up separate BayesDB
+    instances, this function accepts a BayesDB filename, rather than an actual
+    bayeslite.BayesDB object.
+
+    Parameters
+    ----------
+    bdb_file : str
+        File location of the BayesDB database object. This function will
+        handle opening the file with bayeslite.bayesdb_open.
+    table : str
+        Name of the table containing the raw data.
+    model : str
+        Name of the metamodel to estimate from.
+    sim_table : str
+        Name of the table to insert similarity results into. Defaults to
+        table name + '_similarity'.
+    cores : int
+        Number of processors to use. Defaults to the number of cores as
+        identified by multiprocessing.num_cores.
+    N : int
+        Number of rows for which to estimate pairwise similarities (so
+        N^2 calculations are done). Should be used just to test small
+        batches; currently, there is no control over what specific pairwise
+        similarities are estimated with this parameter.
+    overwrite : bool
+        Whether to overwrite the sim_table if it already exists. If
+        overwrite=False and the table exists, function will raise
+        sqlite3.OperationalError. Default True.
+    """
+    bdb = bayesdb_open(pathname=bdb_file)
+
+    if cores is None:
+        cores = mp.cpu_count()
+
+    if cores < 1:
+        raise ValueError(
+            "Invalid number of cores {}".format(cores)
+        )
+
+    if sim_table is None:
+        sim_table = table + '_similarity'
+
+    # Get number of occurrences in the database
+    count_cursor = bdb.execute('SELECT COUNT(*) FROM {}'.format(table))
+    table_count = int(cursor_to_df(count_cursor)['"COUNT"(*)'][0])
+    if N is None:
+        N = table_count
+    elif N > table_count:
+        raise ValueError(
+            "Asked for N={} rows but {} rows in table".format(N, table_count)
+        )
+
+    # Calculate the size (# of similarities to compute) and
+    # offset (where to start) calculation for each worker query.
+
+    # Divide sizes evenly, and make the last job finish the remainder
+    sizes = [(N * N) / cores for i in range(cores)]
+    sizes[-1] += (N * N) % cores
+
+    total = 0
+    offsets = [total]
+    for size in sizes[:-1]:
+        total += size
+        offsets.append(total)
+
+    q_template = ('ESTIMATE SIMILARITY FROM PAIRWISE {} '.format(model) +
+                  'LIMIT {} OFFSET {}')  # Format sizes/offsets later
+
+    queries = [q_template.format(*so) for so in zip(sizes, offsets)]
+
+    # Create the similarity table. Assumes original table has rowid column.
+    # XXX: tables from verbnet bdb don't necessarily have an
+    # autoincrementing primary key other than rowid (doesn't work).
+    # So we ought to ask for a foreign key, but ESTIMATE SIMILARITY
+    # returns numerical values rather than row names, so that code
```

Comment:
To clarify, the issue was with linking the similarity table to the original table via a foreign key. Sqlite3 doesn't accept the autoincrementing rowid as a foreign key (see [here](http://stackoverflow.com/questions/17002251/why-does-referencing-a-sqlite-rowid-cause-foreign-key-mismatch)). So it might be possible for the user to specify a foreign key that's part of the original table, but since Bayeslite's `ESTIMATE SIMILARITY` identifies similarities by the numeric rowid, i.e.

    rowid0 | rowid1 | value
         0 |      0 | 1.000

then there would have to be some kludging to get the rows labeled by the foreign key instead, e.g. (for verbnet)

    rowid0 | rowid1 | value
       run |    hit | 1.000

I can implement that kludging, of course. But initially I simply left off the foreign key constraint.